### PR TITLE
[rbac-manager] Support workload identity

### PR DESCRIPTION
**Pull request by jace-ys**
Hey folks 👋🏻

Hope you don't mind this contribution but we'd like to see theatre support workload identity in the `rbac-manager` instead of using service account keys. I've made the change such that if workload identity is not configured, the `rbac-manager` will fallback to using service account keys.

This is how we're currently using it with workload identity in our GKE cluster (after removing `GOOGLE_APPLICATION_CREDENTIALS`):

Same change on our fork: https://github.com/duffelhq/theatre/pull/3

```yaml
# Config Connector CRDs
---
apiVersion: iam.cnrm.cloud.google.com/v1beta1
kind: IAMPolicy
metadata:
  name: theatre-workload-identity-user
  annotations:
    cnrm.cloud.google.com/project-id: duffel-prod
spec:
  bindings:
  - members:
    - serviceAccount:duffel-prod.svc.id.goog[theatre-system/theatre-rbac-manager]
    role: roles/iam.workloadIdentityUser
  - members:
    - serviceAccount:theatre@duffel-prod.iam.gserviceaccount.com
    # Required so that the theatre service account can impersonate itself
    role: roles/iam.serviceAccountTokenCreator
  resourceRef:
    apiVersion: iam.cnrm.cloud.google.com/v1beta1
    kind: IAMServiceAccount
    external: projects/duffel-prod/serviceAccounts/theatre@duffel-prod.iam.gserviceaccount.com
```

```sh
 kubectl annotate serviceaccount theatre-rbac-manager \
    --namespace theatre-system \
    iam.gke.io/gcp-service-account=theatre@duffel-prod.iam.gserviceaccount.com
```

### CODE REVIEW

**Comment by jace-ys** [comment]
@vinayvinay I think you're the one left in GC that I know..

Any idea who would be best suited to review this? 😁

File: Makefile [hunk 1]
```diff
@@ -68,7 +68,7 @@ manifests: generate
 install-tools:
 	go install github.com/onsi/ginkgo/ginkgo@v1.16.5
 	go install sigs.k8s.io/controller-tools/cmd/controller-gen@v0.10.0
-	go install sigs.k8s.io/controller-runtime/tools/setup-envtest@latest
+	go install sigs.k8s.io/controller-runtime/tools/setup-envtest@release-0.17
```
Comment (jace-ys) [inline Makefile:71]: Shamelessly stolen from #335 to get my CI tests passing.. 😬

**Comment by jace-ys** [comment]
@mbfisher @bogvak @0x0013 Sorry for the direct tag but wanted to get this merged and saw that you folks have been recently active on this repo!

**Comment by mbfisher** [comment]
> @mbfisher @bogvak @0x0013 Sorry for the direct tag but wanted to get this merged and saw that you folks have been recently active on this repo!

I'm a contributor so can't help I'm afraid, Core Infra own this so I'd raise a SPOC ticket with them to get a review

**Comment by jace-ys** [comment]
> > @mbfisher @bogvak @0x0013 Sorry for the direct tag but wanted to get this merged and saw that you folks have been recently active on this repo!
> 
> I'm a contributor so can't help I'm afraid, Core Infra own this so I'd raise a SPOC ticket with them to get a review

Thanks for letting me know! Unfortunately I'm no longer at GC so don't think I'd be able to do that 😅

**Comment by benwh** [comment]
Bumping - it'd be great to get this in!

👋  @ttamimi @VSpike @ijames-gc

File: cmd/rbac-manager/main.go [hunk 2]
```diff
@@ -110,6 +112,30 @@ func createGoogleDirectory(ctx context.Context, subject string) (*directoryv1.Se
...
+		}
+
+		// Access to the directory API must be signed with a Subject to enable domain selection.
+		config := impersonate.CredentialsConfig{
+			TargetPrincipal: principal,
+			Scopes:          scopes,
+			Subject:         subject,
+		}
+
+		ts, err := impersonate.CredentialsTokenSource(ctx, config, option.WithCredentials(creds))
```
Comment (0x0013) [inline cmd/rbac-manager/main.go:131]: Is it necessary for the account to impersonate itself, instead of using the TokenSource already returned by `FindDefaultCredentials`?

File: cmd/rbac-manager/main.go [hunk 3]
```diff
@@ -110,6 +112,30 @@ func createGoogleDirectory(ctx context.Context, subject string) (*directoryv1.Se
...
+		if err != nil {
+			return nil, err
+		}
+
+		return directoryv1.NewService(ctx, option.WithTokenSource(ts))
```
Comment (0x0013) [inline cmd/rbac-manager/main.go:136]: I wonder if the option when using a supplied json credentials (old behavior) could also use the TokenSource from `Credentials` returned by `FindDefaultCredentials`?

If so, likely the code could be simplified a little, instead of needing to call `NewService` with different option for each credential type.

**Review by 0x0013 — COMMENTED** [review]
@jace-ys Thanks, this looks like a valuable addition.

I'm no expert in GCP authentication flow, so I would like to ask some clarifications before I can review, mostly for my own understanding. I've left these in code comments. Apologies if the answers to these are obvious.

File: cmd/rbac-manager/main.go (hunk 3 above)
Comment (jace-ys) [reply cmd/rbac-manager/main.go]: Yeap that's a good point! I didn't change that initially as I didn't want to make too much changes to existing code.

But you're (mostly) right that we can use the same `WithTokenSource` option for both the new and old behaviours. However, it's not `TokenSource` from the `Credentials` returned by `FindDefaultCredentials` but the `TokenSource` from the `jwt.Config`, as we need the token source to have the (domain-wide delegation) subject set.

Comment (jace-ys) [reply cmd/rbac-manager/main.go]: I've updated the PR to use WithTokenSource for both codepaths 👍🏻

File: cmd/rbac-manager/main.go (hunk 2 above)
Comment (jace-ys) [reply cmd/rbac-manager/main.go]: Good question. I'm no expert in GCP auth myself and it's not very well documented or discussed anywhere, but from reading various sources, my understanding is that the federated access token obtained from the GCE
metadata server (via workload identity) is not sufficient for acting as the subject via domain-wide delegation to call Google Workspace Admin APIs.

For delegation to work, we need to sign a JWT with the the "sub" claim set to subject - this happens implicitly through impersonation.

You can kind of this behaviour in `userTokenSource`: https://github.com/googleapis/google-api-go-client/blob/af0a938ba1e01e7c2ce7f96fef1e62478f5da784/impersonate/user.go#L96-L103

Whereas you don't see this in `computeSource` obtained by WID: https://cs.opensource.google/go/x/oauth2/+/refs/tags/v0.24.0:google/google.go;l=270

Comment (jace-ys) [reply cmd/rbac-manager/main.go]: Also, the changes in this PR were largely inspired / adapted from https://github.com/dexidp/dex/pull/2989

Comment (jace-ys) [reply cmd/rbac-manager/main.go]: Some useful reading:
- https://blog.salrashid.dev/articles/2021/impersonation_and_domain_delegation
- https://cloud.google.com/iam/docs/workload-identity-federation#access_management

Comment (jace-ys) [reply cmd/rbac-manager/main.go]: That said, looks like there have been some new developments since I first opened this PR.. 👀 

It looks like it's possible to assign Workspace Admin `Groups Reader` role to service accounts:
- https://support.google.com/a/answer/9807615?sjid=13011041076049488223-AP#:~:text=Assign%20a%20role%20to%20a%20service%20account
- https://cloud.google.com/identity/docs/how-to/setup#auth-no-dwd

Which might mean that there's no need for impersonation or domain-wide delegation anymore, which would greatly simplify the auth here (just `google.FindDefaultCredentials` might be all that's needed) 🤔

But I think that warrants a separate PR as it's a different feature altogether..

**Comment by 0x0013** [comment]
@jace-ys it seems the container-builder hasn't built an image for this branch - perhaps it didn't receive the webhook at the time of your last push.

Could you generate a push event on this PR, for example, by pushing an empty commit, to see if it triggers a build? Thanks!

**Comment by jace-ys** [comment]
> @jace-ys it seems the container-builder hasn't built an image for this branch - perhaps it didn't receive the webhook at the time of your last push.
> 
> Could you generate a push event on this PR, for example, by pushing an empty commit, to see if it triggers a build? Thanks!

Tried that but doesn't look like it triggered a build either! 🙁

**Comment by 0x0013** [comment]
> Tried that but doesn't look like it triggered a build either! 🙁

@jace-ys thanks for trying, it looks like container-builder is still not picking it up. I assume this is because the head branch is external from GC.

I'll see if there's anything that can be done about it.

**Comment by 0x0013** [comment]
Hi @jace-ys ,

Thanks for reminder. It's been a bit hard to commit time to it, but we do have a fix for this issue, just awaiting proper review.

TLDR is that our container-builder, which also is (and should be) a required CI check, currently only acts on organisation-internal commits - which is good from the point of security (we don't want to run automated builds on arbitrary commits), but is obviously not ideal for accepting external contributions. So I've been trying to implement a way to manually trigger it after someone from GC evaluates the changes in the PR. Hopefully we have the reviews done this week.

**Comment by 0x0013** [comment]
Looks like it doesn't work yet, will investigate.

**Comment by jace-ys** [comment]
> Hi @jace-ys ,
> 
> 
> 
> Thanks for reminder. It's been a bit hard to commit time to it, but we do have a fix for this issue, just awaiting proper review.
> 
> 
> 
> TLDR is that our container-builder, which also is (and should be) a required CI check, currently only acts on organisation-internal commits - which is good from the point of security (we don't want to run automated builds on arbitrary commits), but is obviously not ideal for accepting external contributions. So I've been trying to implement a way to manually trigger it after someone from GC evaluates the changes in the PR. Hopefully we have the reviews done this week.

Thanks for looking into this, appreciate your time!

**Comment by 0x0013** [comment]
@jace-ys looks like container-builder is finally able to build the container and also return the status to GitHub CI check.

Could you please rebase this PR to `master` so we can test the resulting container image?

**Review by 0x0013 — DISMISSED** [review]
Sorry this took so long to review due to the CI check issues, and thanks for addressing my previous comments.

The change looks good to me. I've tested on lab to ensure the original behavior (mounted credentials file) still works, and it seems that it does.

Thanks again for submitting the PR.

**Comment by 0x0013** [comment]
@jace-ys it looks like there are some merge conflicts since the dependency updates we pushed yesterday. Could I ask you to rebase once again?

**Review by 0x0013 — APPROVED** [review]
Thanks for rebasing. Looks good.

And thanks again for your contribution.

**Comment by jace-ys** [comment]
@0x0013 Side question, I noticed that the theatre image is still pointing to `gcr.io` when Google have deprecated Container Registry for Artifact Registry.

So wanted to know if GC have redirected all `gcr.io` to Artifact Registry, or should we publish our own image to Artifact Registry?
